DM-27926: Draft Lightweight/Hybrid alert technote

ebellm opened the pull request.

ebellm committed:
initial draft

ktlim commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+\citeds{DMTN-093}.
+As the Alert Production pipeline runs, it would produce both the full-sized alert packets and their lightweight counterparts, both in Avro format.
+The lightweight alerts would be distributed to community brokers and science users using Kafka.
+The full-sized alerts would be stored on disk in an object store or other system accessible by HTTP.
```
Although I can see how to make this work with HTTP, please consider using HTTPS.  Even though the Alert stream is world-public, we may need authentication to constrain bandwidth.  HTTPS can also ensure integrity (although I believe we will still want to sign lightweight and full-sized alerts).

ktlim commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+
+%  URL for packet retrieval -- longevity
+The link to retrieve the full-sized alert must be permanent, even if the full-sized alert is moved.  
+This suggests use of an identifier such as a DOI, URI, or IVORN\footnote{\url{https://www.ivoa.net/documents/IVOAIdentifiers/20160523/index.html}}.
```
Indirection, particularly through a third party, may add substantial latency for both publication/availability as well as each retrieval.

ktlim commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+Particularly if flexible identifiers pointing to the full-sized alerts are used, it seems useful to investigate whether the Alert Database and the store of full-sized alerts could be the same system, technically.
+This would provide a seamless transition for the user between active and archival alerts.
+Users could retrieve full-sized alerts by their URIs, and it would be straightforward to implement additional queries (cone search, time window, etc.) by querying the Prompt Products Database and retrieving the appropriate alert URIs.
+It would also be necessary to archive the lightweight alerts, but due to their small size this should not be a large perturbation.
```
If the lightweight alerts are a strict subset of the full-sized, they could be regenerated dynamically if needed.

leannep replied:
I don't see the archiving of the lightweight alerts represented in figure 2, just the full alerts.

ebellm replied:
@leannep There's an arrow coming out of "Alert Packaging" to the Alert Database that I think can count for archiving the lightweight alerts.

ktlim commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+
+%need for authentication (for data rights, also for rate limits)
+Both the lightweight alert stream and the packet retrieval service would require authentication and authorization to confirm data rights and set appropriate rate limits.
+Individuals may find rate limits frustrating if their alert retrieval needs don't match their rate limits and may pressure the project for greater access rather than adopting their workflow to use a community broker.
```
```suggestion
Individuals may find rate limits frustrating if their alert retrieval needs don't match their rate limits and may pressure the project for greater access rather than adapting their workflow to use a community broker.
```

ktlim commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+
+The lightweight alerts contain a URI that points to the corresponding full alert packet.
+If a user without data rights is unable to access the packet at that location, having a copy of the world-public lightweight alert is not very useful.
+While community brokers are responsible for providing public access to alerts even in the baseline, they would need to provide their own additional URL or URI pointing to a publicly-accessible copy of the full alert, which could create confusion.
```
This can be a broker-hosted service at a different endpoint that accepts the same URI string; it doesn't require a new URI per alert.

ktlim commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+From a pure user-experience perspective this suggests making access to the full alerts world-public, even if the rate limits for non-data-rights holders are extremely low.
+
+Referential integrity of the lightweight and full alerts could be hard to maintain. 
+Perfect synchronization of the data and notification systems is probably impossible, which could impact users' ability to rapidly retrieve the full alerts.
```
Perfect synchronization may be difficult, but ensuring that the full alert is present before publishing a lightweight alert is somewhat easier to guarantee (i.e. the lightweight stream may miss some but may not publish extra).

ktlim commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+totFluxSigma & 	float	& 4 \\
+distance to nearest star & 	float	& 4 \\
+distance to nearest galaxy	& float	& 4 \\
+parallax/PM	& 3 floats &	9 \\
```
```suggestion
parallax/PM	& 3 floats &	12 \\
```

ivezic commented:
I read the version from 2020-12-09. Excellent work, it's pretty much ready for circulation. Given that somewhat arbitrary selection of fields in Table 1 gives ~200 bytes/alert, and leads to a ~500 times smaller size, I wonder if a less radical proposal to include the entire DIAObject (1.4 kB) and the DIASource which triggered the alert (0.45 kB) and thus have light-weight alerts of ~2 kB would still be acceptable (i.e. ~50 times smaller size). Or perhaps have both ("lightweight" and "middleweight" as in boxing) and the user can choose which one to subscribe too (with some preset limits)?

ctslater commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+%    Peaky usage of bandwidth
+%    Kafka doesn't like large messages
+
+Our overarching goal for considering a hybrid alert architecture is the enhance the scientific return of the LSST alert stream by ensuring a large and robust broker and user ecosystem and by improving performance and access for individual users relative to the baseline.
```
```suggestion
Our overarching goal for considering a hybrid alert architecture is to enhance the scientific return of the LSST alert stream by ensuring a large and robust broker and user ecosystem and by improving performance and access for individual users relative to the baseline.
```

MelissaGraham commented:
This is excellent and an exciting option -- I read through and encountered just a few questions.

Figure 1 and 2 have arrows from the data rights community into the PPDB, DRDB, and ADB which gives the impression of the community inputting to those containers but that's probably an accident?

Typo section 2 paragraph 2 sentence 1 "the enhance".

Section 9 paragraph 4 sentence 1 "to retrieve the full alert packet imposes a small amount of additional latency (in the best case, roughly 200 msec)", two questions: (1) should it be "packets" b/c it is not 200 msec per packet, and (2) is there a worst-case scenario that could/should be quoted.

Section 9 paragraph 4 middle, is the gain for the broker in Rubin being able to distribute the alerts more broadly that the broker would not have to support as big a community of users? I'm not sure they would see that as a gain, but maybe I guess.

Section 9 paragraph 5 middle, the part that said "only users with data rights could subscribe to the lightweight stream" might need a bit further clarification because the brokers themselves need not have data rights. Maybe it should say that the full broker application process does not require the broker developers to have data rights, but individual access to the lightweight stream would be via the Rubin Data Facility so thus would require data rights? Or whatever it is most appropriate to say.

ctslater commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+Most importantly, it makes outbound bandwidth from the Data Facility the primary constraint on the system.
+As a result, only a small number of consumers can receive the full alert stream.
+This limits the reach and impact of the alert stream and hinders the flexibility and creativity of scientists.
+Moreover, both brokers and users must stream through each large packet on their system, even if they intend to filter it out and discard it---this makes handling the alert stream more technically challenging than is necessary.
```
"must stream through" is a sort of vague complaint. I can understand a general notion that working with less data is easier than more data, but is there a more precise way to describe this issue? (I would also avoid the word stream)

ebellm replied:
Does "seek through" make it more clear?  We are transmitting each packet with the Confluent Wire Format wrapping a [single object](https://avro.apache.org/docs/current/spec.html#single_object_encoding) Avro encoded packet.  So I think Kafka has to touch all the bytes to find the next message.

ctslater commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+
+Our overarching goal for considering a hybrid alert architecture is the enhance the scientific return of the LSST alert stream by ensuring a large and robust broker and user ecosystem and by improving performance and access for individual users relative to the baseline.
+At the 2019 community broker workshop, there was clear consensus among the attendees that the Project should explore avenues that would broaden access to the alert stream.
+At the same time we want to maintain the rich content of the alert packet to enable rapid filtering and analysis and enable the alerts to be widely shared publicly.
```
I think the rich alert content is to enable science, not filtering. (the main argument of the document is that lightweight alerts are sufficient for filtering.)

ctslater commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+Before describing the hybrid alert concept, we first stress that we are not advocating for a change to the relevant requirements\lsrreq{0101}\reqparam{OTT1}\dmreq{0391}\reqparam{numStreams}\dmreq{0274}.
+We will still produce all of the required alert contents and be capable of transmitting them to the required number of brokers within the required 60 second window after readout completes.
+
+In the hybrid alert packet concept, the realtime alert stream still contains information about all detection \DIASources.
```
```suggestion
In the hybrid alert packet concept, the realtime alert stream still contains information about all detected \DIASources.
```

ctslater commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+In contrast to the full alert stream, which is $\sim$ 820\,GB per night, it is easy to imagine a graduate student accessing the full alert stream from their laptop, or perhaps connecting a small or ad-hoc automated telescope to it.
+We believe that democratizing access to the alert stream will enable some of the most creative uses of the stream and ensure innovative use by scientists.
+
+Each lightweight alert packet would include a link to allow the user to retrieve the full alert packet containing all of the DPDD-specified quantities (image cutouts, \DIASource history, forced photometry where available, timeseries features, etc.).
```
I'd consider moving this paragraph up to immediately follow the "In the hybrid alert packet concept" paragraph, since it's the first description of the 'other half' of the hybrid.

ctslater commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+Additionally it requires brokers to re-engineer existing alert retrieval code.
+The handful of full-stream brokers may view these as pure costs with no benefit relative to the baseline.
+We believe these costs are justified by the gains of being able to distribute the alerts more widely; the modest additional latency is not scientifically significant.
+As discussed above, brokers may ultimately find it technically advantageous to be able to have their own systems direct and parallelize the retrieval of the largest volumes of data.
```
Not sure if it fits in here, but I expect a variety of catch-up, resync and completeness checking scenarios to be a lot easier for the brokers this way.

ebellm replied:
This is a good point--this comes up frequently in the ZTF context, actually.

ctslater commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+The AFS is likely to have had hard limits at 20 full-sized alerts per visit, after which no more would be forwarded.
+In contrast, a user wanting many more alerts from a high-priority single visit (say, a Deep Drilling field) could simply request more full-sized alerts over a longer period of time.
+
+Descoping the AFS would reduce on-Project construction and operations effort.
```
I think the alert filtering system is a pretty big technical risk. Descoping it is a big feature of the proposal to me.

leannep replied:
Suggest:  "would reduce both effort and risk for the  construction project and operations effort."

ctslater reviewed (commented):
Looks really good.

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+Alerts are sent to third-party community brokers for further enhancement and redistribution and are also available to Data Rights holders through the Alert Filtering Service.
+
+In order to provide all of the information needed to make rapid classification and follow-up decisions, the alert packets are ``rich''--they contain not only the information about the latest detection, but also past detection history, forced photometry and/or upper limits, the associated \DIAObject or \SSObject record, linkages to counterparts in the most recent LSST data release, and image cutou ...
+The alert contents are public and freely shareable \citedsp{RDO-013}.
```
rdo-013 doesn't render in the latex - it is in docushare but needs to be added to the bib files

leannep replied:
(I added RDO-013 to lsst.bib)

ebellm replied:
It shouldn't be necessary--Wil put it in the lsst-texmf yesterday.

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+Instead of providing filter code to run on Data Facility services, users with data rights could run filters on their own hardware with a great deal more flexibility.
+In particular, users would be free to build filters that crossmatched to external catalogs or realtime streams or performed computationally-intensive fitting or image-processing tasks.
+They would be free to choose their own programming language, libraries, and environment.
+While the need to implement their own local service would impose additional difficulty relative to a Project-provided AFS, the Project could provide simple tutorial material for a basic implementation.
```
I'd make this stronger - we must provide material to help users set up their own service if we descope the AFS

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+
+%\textbf{estimate data rates for 100x20.  20 full-sized alerts per filter had to come out anyway}
+
+One particular advantage of this scheme to users is that within their overall rate limit budget, they could receive more alerts from single visits at cost of higher latency.
```
they could receive more alerts from single visits at **the** cost of higher latency.

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+Bulk transport and rapid notification workloads tend to have quite different architectures.
+
+%  More effective use of datacenter bandwidth
+While we will still meet the 60-second latency requirement, in practice we expect that this design will produce more effective use of the datacenter bandwidth.
```
Suggest:  this design will **result in a** more effective use of the datacenter bandwidth.

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+While not a feature of the design presented here, we speculate that users may wish to retrieve only a portion of the full-sized alert---the cutouts, say, or the \DIASource history.
+A small service extracting only the relevant alert portion would provide further improvements in bandwidth use.
+
+Finally, the proposed system provides potential improvements in interfacing with the alert filtering service and alert archive, as discussed in \S \ref{sec:alertfiltering} and \S \ref{sec:alertdb}.
```
You have advocated for descoping the AFS in section 5. "We suggest that the lightweight alert packet design would obviate the scientific need for the AFS, reducing Project scope" . This statement now seems contradictory.

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+\section{Challenges and Concerns}
+%  Two systems, not one (but: may save labor with alertdb/lafs)
+Among the technical challenges presented by the hybrid alert design is that it requires the Rubin Observatory project to design and administer two systems (the lightweight stream and the full-size alert retrieval system) rather than one (a full-sized alert stream).
+Given the late stage of the construction project it is reasonable to question whether it is prudent to consider such a change.
```
maybe add that the alert production system **must** be ready for the start of operations to support early science.

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+While moving to smaller alerts will improve Kafka's performance, distributing a lightweight Kafka stream to a much wider audience may require further attention.
+We do not anticipate challenges scaling the Kafka system itself to many consumers, but Kafka is primarily used as a streaming solution \textit{within} organizations, and as of this writing its capabilities for handling many untrusted third-party consumers is less mature.
+
+It will require effort and scientific consensus-building to identify a minimal subset of alert contents for lightweight alerts that maintains its small size but provides the necessary information for efficient filtering.
```
Suggest: "consensus-building with the Rubin community". Th Rubin community is: "a broad term that refers to anyone, anywhere, interacting with the Rubin Observatory and its data products and services, in any capacity. It is a union set of all of the following communities, and is more than just data rights holders."

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+
+It will require effort and scientific consensus-building to identify a minimal subset of alert contents for lightweight alerts that maintains its small size but provides the necessary information for efficient filtering.
+There will be community demand to repeatedly increase the number of included fields to satisfy individual science cases.
+
```
Another concern I would add is the possible increased burden of support on operations. Rather than going only through community brokers,  the community who directly access the stream will come to the project for help.  Brokers are currently a 3rd party community activity. The new architecture transfers some of that responsibility to the project/operations.

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+A broad concern is whether lightweight alerts are beneficial to community brokers.
+We believe from a technical standpoint that brokers will benefit from being able to retrieve the full-size alerts when desired and in parallel, rather than being forced to stream through each alert to get to the next one.
+But for the top five selected brokers this benefit is probably not large.
+The greater advantage is to community brokers that might not otherwise be able to access the full alert stream, either directly from the project or from an ``upstream'' community broker.
```
Suggest: "not otherwise be able to **or want to** .. " Several broker LOIs were science specific or targeted in some way.

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,289 @@
+
+% Is this just a broker function?
+As discussed in \S \ref{sec:alertfiltering}, the biggest beneficiaries of the lightweight alert stream are likely to be single science users or small groups, who can now receive the complete lightweight alert stream and perform more sophisticated analysis than is possible with the baselined Alert Filtering Service.
+It may be argued that filtering and/or redistribution of alerts with fewer contents can or will be performed by the brokers themselves, and Rubin's efforts are thus duplicative (and may even hinder broker teams' ability to obtain funding).
```
Can or may be but they are not obliged to and could they could stop such a redistribution at any time (or lose finding to do so)

leannep reviewed (changes requested):
Really nice writeup. The scientific and technical motivations are well laid out and convincing.

leannep commented:
> I read the version from 2020-12-09. Excellent work, it's pretty much ready for circulation. Given that somewhat arbitrary selection of fields in Table 1 gives ~200 bytes/alert, and leads to a ~500 times smaller size, I wonder if a less radical proposal to include the entire DIAObject (1.4 kB) and the DIASource which triggered the alert (0.45 kB) and thus have light-weight alerts of ~2 kB would still be acceptable (i.e. ~50 times smaller size). Or perhaps have both ("lightweight" and "middleweight" as in boxing) and the user can choose which one to subscribe too (with some preset limits)?

leannep commented:
One option to consider is to still send the full stream to the general purpose brokers for whom this lightweight model brings fewer benefits and to insist that all non data rights holders access the public alerts via a community broker.  The lightweight model could then be reserved for data rights holders including specialized brokers.

ebellm commented:
Replying to @MelissaGraham 

> Figure 1 and 2 have arrows from the data rights community into the PPDB, DRDB, and ADB which gives the impression of the community inputting to those containers but that's probably an accident?

This was meant represent queries back to those dbs--would arrows both directions be better?
 
> Typo section 2 paragraph 2 sentence 1 "the enhance".

Fixed.

> Section 9 paragraph 4 sentence 1 "to retrieve the full alert packet imposes a small amount of additional latency (in the best case, roughly 200 msec)", two questions: (1) should it be "packets" b/c it is not 200 msec per packet, and (2) is there a worst-case scenario that could/should be quoted.

Sorry, it's not the roundtrip to transmit the packet, it's some assumed network travel time for the query itself.  I'd have to get Spencer to expand on this when he returns.

> Section 9 paragraph 4 middle, is the gain for the broker in Rubin being able to distribute the alerts more broadly that the broker would not have to support as big a community of users? I'm not sure they would see that as a gain, but maybe I guess.

I think I'll leave this out so as not to antagonize any brokers...

> Section 9 paragraph 5 middle, the part that said "only users with data rights could subscribe to the lightweight stream" might need a bit further clarification because the brokers themselves need not have data rights. Maybe it should say that the full broker application process does not require the broker developers to have data rights, but individual access to the lightweight stream would be via the Rubin Data Facility so thus would require data rights? Or whatever it is most appropriate to say.

I added "and selected brokers"

ebellm committed:
response to comments

leannep commented on DMTN-165.tex:
```diff
@@ -50,7 +56,302 @@
+Ensuring that the full alert is present for retrieval before publishing the lightweight alert could be guaranteed
+but may impact users' ability to rapidly retrieve the full alerts.
+
+Because brokers must now actively retrieve the full alert packets, this inhibits arch/itectures where the brokers provide only stateless real-time filtering or forwarding of alerts.
```
arch/itectures -> architectures

leannep reviewed (approved):
One minor typo only; good to go
